source/code/providers/support/mycnf-scanner.cpp

```diff
@@ -0,0 +1,244 @@
+/*
+ * --------------------------------- START OF LICENSE ----------------------------
+ *
+ * MySQL cimprov ver. 1.0
+ *
+ * Copyright (c) Microsoft Corporation
+ *
+ * All rights reserved. 
+ *
+ * MIT License
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the ""Software""), to deal
+ * in the Software without restriction, including without limitation the rights to
+ * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies
+ * of the Software, and to permit persons to whom the Software is furnished to do
+ * so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED *AS IS*, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ * ---------------------------------- END OF LICENSE -----------------------------
+*/
+/**
+    \file        mycnf-scanner.cpp
+
+    \brief       Scans mysql configuration to extract authentication information.
+
+    \date        2014-11-05 16:22:00
+
+*/
+/*----------------------------------------------------------------------------*/
+
+#include "mycnf-scanner.h"
+
+namespace MySQL
+{
+    static inline bool GetPathFromDirective(const std::wstring& line, SCXCoreLib::SCXFilePath& path)
+    {
+        std::vector<std::wstring> elements;
+        SCXCoreLib::StrTokenize(line, elements, L" ");
+        if ( elements.size() != 2 )
+        {
+            return false;
+        }
+
+        path.Set(elements[1]);
+        return true;
+    }
+
+    bool MySQL_ConfigurationScanner::LoadCnfFile(
+        const SCXCoreLib::SCXFilePath& file, 
+        MySQL_Authentication& auth,
+        std::set< std::wstring >& visited, 
+        std::map<unsigned int, std::wstring>& ports)
```

Comment:
I don't see a use-case for this. By keeping state inside of the object, you "hide" complexity (by having fewer parameters), and you allow for the Scan method to change implementations. Right now, the implementation is somewhat fixed because of what you're passing for state.

Besides, setting up logging is non-trivial (getting the m_log handle, etc). It's better to have the constructor take care of that once for the life of the object.

If there was a compelling use case, you could make this a static member and pass state. But, so far, we never do that unless there's a compelling need. I don't see a compelling need here, and I see reasons to not do that (performance, etc). In most cases, we do not do static classes unless, for some reason, we really need that. And, often, that makes maintenance of the class harder if we need to make changes later (like I just did recently to the PAL networking class).

I would strongly recommend changing the class to take care of it's own state, thus hiding implementation within the class. This is basic Object Design, and makes for more maintainable code.

---
In reply to: [66896591](https://github.com/Microsoft/MySQL-Provider/pull/5#discussion_r66896591) [](ancestors = 66896591)